presto-main/src/main/java/io/prestosql/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -419,7 +420,7 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
         }
         Set<SqlStageExecution> childStages = childStagesBuilder.build();
         stage.addStateChangeListener(newState -> {
-            if (newState.isDone()) {
+            if (newState.isDone() || newState == FLUSHING) {
```

Comment:
Please add a test for this. The test could be:
1. staring `SELECT * FROM large_table LIMIT 1`
2. NOT fetching results and waiting for SOURCE stage to be cancelled